### PR TITLE
openapi(dm): support show task status in list task API

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

Issue Number: ref #3583

### What is changed and how it works?

as title

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation

### Release note 

```release-note
`support show task status in DM list task OpenAPI`.
```

### CODE REVIEW

File: dm/dm/master/openapi_test.go [hunk 1]
```diff
@@ -712,7 +728,7 @@ func mockTaskQueryStatus(
 	}
 	mockWorkerClient.EXPECT().QueryStatus(
 		gomock.Any(),
-		&pb.QueryStatusRequest{Name: taskName},
```
Comment (Ehco1996) [inline dm/dm/master/openapi_test.go:715]: revert this

File: dm/dm/master/openapi.go [hunk 2]
```diff
@@ -840,3 +803,74 @@ func terrorHTTPErrorHandler() gin.HandlerFunc {
 		c.IndentedJSON(http.StatusBadRequest, openapi.ErrorWithMessage{ErrorMsg: msg, ErrorCode: code})
 	}
 }
+
+func getOpenAPISubtaskStatusByTaskName(taskName string,
```
Comment (Ehco1996) [inline dm/dm/master/openapi.go:807]: this is moved for L550 (get task status)

**Comment by Ehco1996** [comment]
could you please take another look for this pr ? thanks 🧡🧡🧡

@glorv  @okJiang @GMHDBJD

**Review by okJiang — APPROVED** [review]

File: dm/dm/master/openapi.go [hunk 3]
```diff
@@ -521,10 +521,36 @@ func (s *Server) DMAPIDeleteTask(c *gin.Context, taskName string, params openapi
...
+	// fill status
+	if params.WithStatus != nil && *params.WithStatus {
+		// get source list for all task
+		var sourceList []string
+		sourceNameM := make(map[string]struct{})
+		for _, task := range taskList {
+			for _, sourceConf := range task.SourceConfig.SourceConf {
+				sourceNameM[sourceConf.SourceName] = struct{}{}
+			}
+			for sourceName := range sourceNameM {
```
Comment (lance6716) [inline dm/dm/master/openapi.go:538]: this should be moved to outer level?

File: dm/dm/master/openapi.go [hunk 4]
```diff
@@ -521,10 +521,36 @@ func (s *Server) DMAPIDeleteTask(c *gin.Context, taskName string, params openapi
...
+	if params.WithStatus != nil && *params.WithStatus {
+		// get source list for all task
+		var sourceList []string
+		sourceNameM := make(map[string]struct{}) // use map to avoid duplicate source name
+		for _, task := range taskList {
+			for _, sourceConf := range task.SourceConfig.SourceConf {
+				sourceNameM[sourceConf.SourceName] = struct{}{}
+			}
+		}
+		for sourceName := range sourceNameM {
```
Comment (lance6716) [inline dm/dm/master/openapi.go:539]: maybe utils.SetToSlice

File: dm/dm/master/openapi.go [hunk 5]
```diff
@@ -521,10 +521,36 @@ func (s *Server) DMAPIDeleteTask(c *gin.Context, taskName string, params openapi
...
+		// get status from workers
+		workerStatusList := s.getStatusFromWorkers(c.Request.Context(), sourceList, "", false)
+		// fill status for every task
+		for idx := range taskList {
+			subTaskStatusList, err := getOpenAPISubtaskStatusByTaskName(taskList[idx].Name, workerStatusList)
+			if err != nil {
+				_ = c.Error(err)
+				return
+			}
+			taskList[idx].StatusList = &subTaskStatusList
```
Comment (lance6716) [inline dm/dm/master/openapi.go:551]: please add a integration test for this feature when we have started multiple tasks.

I started `test-shard`, `test`, `testAA` and now the number of `status_list` is not expected

```

            "status_list": [
                {
                    "load_status": {
                        "finished_bytes": 0,
                        "meta_binlog": "(mysql-bin.000001, 2568)",
                        "meta_binlog_gtid": "2c43df5b-5eea-11ec-aff5-0242ac110002:1-12",
                        "progress": "0.00 %",
                        "total_bytes": 315
                    },
                    "name": "testAA",
                    "source_name": "mysql-01",
                    "stage": "Paused",
                    "unit": "Load",
                    "unresolved_ddl_lock_id": "",
                    "worker_name": "worker2"
                },
                {
                    "load_status": {
                        "finished_bytes": 0,
                        "meta_binlog": "(mysql-bin.000001, 2520)",
                        "meta_binlog_gtid": "def3d483-5ede-11ec-b145-0242ac110003:1-11",
                        "progress": "0.00 %",
                        "total_bytes": 158
                    },
                    "name": "testAA",
                    "source_name": "mysql-02",
                    "stage": "Paused",
                    "unit": "Load",
                    "unresolved_ddl_lock_id": "",
                    "worker_name": "worker1"
                },
                {
                    "load_status": {
                        "finished_bytes": 0,
                        "meta_binlog": "(mysql-bin.000001, 2520)",
                        "meta_binlog_gtid": "def3d483-5ede-11ec-b145-0242ac110003:1-11",
                        "progress": "0.00 %",
                        "total_bytes": 158
                    },
                    "name": "testAA",
                    "source_name": "mysql-02",
                    "stage": "Paused",
                    "unit": "Load",
                    "unresolved_ddl_lock_id": "",
                    "worker_name": "worker1"
                },
                {
                    "load_status": {
                        "finished_bytes": 0,
                        "meta_binlog": "(mysql-bin.000001, 2520)",
                        "meta_binlog_gtid": "def3d483-5ede-11ec-b145-0242ac110003:1-11",
                        "progress": "0.00 %",
                        "total_bytes": 158
                    },
                    "name": "testAA",
                    "source_name": "mysql-02",
                    "stage": "Paused",
                    "unit": "Load",
                    "unresolved_ddl_lock_id": "",
                    "worker_name": "worker1"
                },
                {
                    "load_status": {
                        "finished_bytes": 0,
                        "meta_binlog": "(mysql-bin.000001, 2568)",
                        "meta_binlog_gtid": "2c43df5b-5eea-11ec-aff5-0242ac110002:1-12",
                        "progress": "0.00 %",
                        "total_bytes": 315
                    },
                    "name": "testAA",
                    "source_name": "mysql-01",
                    "stage": "Paused",
                    "unit": "Load",
                    "unresolved_ddl_lock_id": "",
                    "worker_name": "worker2"
                },
                {
                    "load_status": {
                        "finished_bytes": 0,
                        "meta_binlog": "(mysql-bin.000001, 2568)",
                        "meta_binlog_gtid": "2c43df5b-5eea-11ec-aff5-0242ac110002:1-12",
                        "progress": "0.00 %",
                        "total_bytes": 315
                    },
                    "name": "testAA",
                    "source_name": "mysql-01",
                    "stage": "Paused",
                    "unit": "Load",
                    "unresolved_ddl_lock_id": "",
                    "worker_name": "worker2"
                }
            ],
```

**Review by lance6716 — CHANGES_REQUESTED** [review]

File: dm/dm/master/openapi.go (hunk 5 above)
Comment (Ehco1996) [reply dm/dm/master/openapi.go:548]: sorry could you please provied the task config?  i can't reproduce this case

the it test is added in https://github.com/pingcap/tiflow/pull/3971/commits/b05c993881abf0cd256c05f3b77575814c271ce6

Comment (lance6716) [reply dm/dm/master/openapi.go:548]: can't reproduce as well :joy:

**Review by lance6716 — APPROVED** [review]

**Pull request merged**